src/metabase/models/card.clj

```diff
@@ -306,27 +339,58 @@
                            :query-database        query-db-id
                            :field-filter-database field-db-id})))))))
 
-(defn- assert-valid-model
+(defn- ensure-type-and-dataset-are-consistent
+  "We're in the process of migrating from using `report_card.dataset` to `report_card.type`.
+  In the future we'll drop `dataset` and only use `type`. But for now we need to make sure that both keys are aligned
+  when dealing with cards.
+  - If both keys are present, we make sure `dataset` is true if `type` is `model` else false.
+    This will make a different when we have `metric` type since a boolean can't represent tri-state
+  - If only one key is present, we'll assoc the correct value for the other key."
+  [{:keys [type dataset] :as card}]
+  (cond
+   ;; if none of the 2 keys presents, do nothing
+   (and (nil? type) (nil? dataset))
+   card
+
+   ;; if both type and dataset presents, we prioritize type
+   (and (some? type) (some? dataset))
+   (assoc card :dataset (= type "model"))
```

Comment:
Can you explain why type is prioritized here?

Reply:
because we can only infer dataset from `type` but not the other way around.

Reply:
Got it. I was concerned they could disagree but I think you're assuming here that they already match because it passed the `check-card-type-and-dataset` check.

Reply:
Actually, I'm concerned this can paper over bugs in ways that are difficult to understand. I can imagine some old function (call it `f`) that takes a card with `:dataset` and `:type`, then updates `:dataset` but not `:type`. When the card gets to `ensure-type-and-dataset-are-consistent` we will undo the change that `f` made.

To defend against this possibility I think we need to assert that `:dataset` and `:type` do not contradict each other here with a runtime exception.

Reply:
I don't want to change this function to throw an exception; its primary job is to fill out `type` and `dataset` if one of them is missing.

instead I've added `assert-card-type-and-dataset` and guard it on both `before-insert` and `before-update`, I think that'll help us avoid the cases where `:dataset` and `:type` is inconsistent.